### DISCUSSION TITLE
How to do GET request query parameter database validation which depends on multiple other query params.

### DISCUSSION THREAD

**Discussion by saanpritom** [fastapi/fastapi#6403] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import datetime

app = FastAPI()

@app.get("/clinic/available/doctors/list/")
def get_supplier_available_experts_list_on_clinic(clinic_id: str, supplier_id: str, date: datetime.date):
    """Return the list of the experts for a given supplier to a given clinic on a specific date."""
    return {"experts": []}
```

### Description

On my above code you can see the function **get_supplier_available_experts_list_on_clinic()** takes three URL parameters. One of them is **clinic_id**, one is **supplier_id** and other is a valid **date**. The function does some internal calculations to generated the list of available experts for a specific supplier to a specific clinic on a given date.

The problem is I want to validate the three parameters as follows:

- There must be a valid clinic on the Database for the inputted Clinic ID
- There must be a valid supplier on the Database for the inputted Supplier ID
- There must be a relation between the Clinic and the Supplier on the Database
- There must be a valid schedule for the Clinic and for the Supplier on the Database for the given Date

Currently, I can do these validations checkups on my internal scripts but I want to do it on the parameter level like using some **Dependencies** or other mechanisms.  Is it currently possible with the FastAPI? 

### Operating System

macOS

### FastAPI Version

0.77.1

### Python Version

3.10.2

### Additional Context

Current Database: PostgreSQL 14

**Comment by acidjunk** [comment]
You could probably use a pydantic validator: to do some heavy lifting.
https://pydantic-docs.helpmanual.io/usage/validators/

But considering that 2 of them need a DB (and a valid record) -> it would probably be good to raise 404 error's when the ID don't exist. And 422 when the relation doesn't exist?

So IMO: it also depends on how you want to design your API.

**Comment by aboubacs** [comment]
Validators don't seem like a valid place for your use case. If you don't want to implement your parameter validation logic in your endpoint (I'm not exactly sure why in the first place ?), you could do it with the dependency system. For example, instead of depending on "clinic_id", your endpoint can depend on "clinic", which depends on the database and a clinic_id (and raises an error if the clinic doesn't exist).

Some sample code would be:

```python

def get_db():
    try:
        db = your_function_to_get_a_database_session()
        yield db
    finally:
        db.close()


def get_clinic(db = Depends(get_db), clinic_id: str):
    # Reads the clinic from clinic_id
    # Raises an exception if the clinic doesn't exist

@app.get("/clinic/available/doctors/list/")
def get_supplier_available_experts_list_on_clinic(clinic: Depends(get_clinic), .......)
    .......
```

**Comment by saanpritom** [comment]
Thanks @aboubacs for the solution. I think this will work. I want to do it on the parameter level because i have multiple URL functions which required the same parameters. So, I have to validate the parameter each time. That's why I want a re usable clean solution.

**Comment by saanpritom** [comment]
Thanks @acidjunk for the reply. I have checked the documentation you sent. But I think the solution which @aboubacs provided is more suited to my problem.

**Comment by tiangolo** [comment, marked as answer]
Thanks for the help here everyone! 🍰 

I agree that using a dependency should do the trick. 🤓 

Thanks @saanpritom for coming back to close the issue once you found a solution, that's highly appreciated.

> Sorry for the long delay! 🙈  I wanted to personally address each issue and they piled up through time, but now I'm checking each one in order.
